bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/GsonMessageBodyReader.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.io.rest.core.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyReader;
+
+import com.google.gson.Gson;
+
+/**
+ * A message body reader for JSON using GSON.
+ *
+ * @author Markus Rathgeb - Initial contribution
+ */
+public class GsonMessageBodyReader<T> implements MessageBodyReader<T> {
```

Comment:
NonNullByDefault

Reply:
That's not possible because it's implementing an unannotated non-OH interface so adding that annotation results in many "Illegal redefinition of parameter annotations, inherited method from ... does not constrain this parameter" errors.

Reply:
That‘s one of the bigger problems we have with our null-annotations. All classes implementing non-annotated (external) interfaces cannot be annotated in our code.

Reply:
Yes we should really get those EEAs working (https://github.com/openhab/openhab-core/issues/888). Maybe we have more time for that when this PR is merged. :-)